src/interface/swarm_pack_base.hpp

```diff
@@ -0,0 +1,265 @@
+struct GetDataType {
+  using value = typename Head::data_type;
+  static_assert(std::is_same<value, typename GetDataType<Tail...>::value>::value,
+                "Types must all be the same");
+};
+template <typename T>
+struct GetDataType<T> {
+  using value = typename T::data_type;
+};
+} // namespace
+
+namespace parthenon {
+
+namespace impl {
+template <typename TYPE>
+struct SwarmPackDescriptor {
+  SwarmPackDescriptor(const std::string &swarm_name, const std::vector<std::string> &vars)
+      : swarm_name(swarm_name), vars(vars), identifier(GetIdentifier()) {}
+
+  // Determining if variable pv should be included in SwarmPack
+  bool IncludeVariable(int vidx,
+                       const std::shared_ptr<ParticleVariable<TYPE>> &pv) const {
+    if (vars[vidx] == pv->label()) return true;
+    return false;
+  }
+
+  const std::string swarm_name;
+  const std::vector<std::string> vars;
+  const std::string identifier;
+
+ private:
+  std::string GetIdentifier() const {
+    std::string ident("");
+    for (const auto &var : vars)
+      ident += var;
+    ident += "|swarm_name:";
+    ident += swarm_name;
+    return ident;
+  }
+};
+} // namespace impl
+
+using SwarmPackIdxMap = std::unordered_map<std::string, std::size_t>;
+
+template <typename TYPE>
+class SwarmPackBase {
+ public:
+  SwarmPackBase() = default;
+  virtual ~SwarmPackBase() = default;
+
+  using pack_t = ParArray3D<ParArray1D<TYPE>>;
+  using bounds_t = ParArray3D<int>;
+  using contexts_t = ParArray1D<SwarmDeviceContext>;
+  using max_active_indices_t = ParArray1D<int>;
+  using desc_t = impl::SwarmPackDescriptor<TYPE>;
+
+  // Actually build a `SwarmPackBase` (i.e. create a view of views, fill on host, and
+  // deep copy the view of views to device) from the variables specified in desc contained
+  // from the blocks contained in pmd (which can either be MeshBlockData/MeshData).
+  template <class MBD>
+  static SwarmPackBase<TYPE> Build(MBD *pmd, const SwarmPackDescriptor<TYPE> &desc) {
+    using mbd_t = MeshBlockData<Real>;
+    int nvar = desc.vars.size();
+
+    SwarmPackBase<TYPE> pack;
+    pack.nvar_ = desc.vars.size();
+
+    // Count up the size of the array that is required
+    int max_size = 0;
+    int nblocks = 0;
+    int ndim = 3;
+    std::vector<int> vardims;
+    ForEachBlock(pmd, [&](int b, mbd_t *pmbd) {
+      auto swarm = pmbd->GetSwarm(desc.swarm_name);
+      int size = 0;
+      nblocks++;
+      for (auto &pv : swarm->template GetVariableVector<TYPE>()) {
+        for (int i = 0; i < nvar; ++i) {
+          if (desc.IncludeVariable(i, pv)) {
+            size += pv->GetDim(6) * pv->GetDim(5) * pv->GetDim(4) * pv->GetDim(3) *
+                    pv->GetDim(2);
+            ndim = 1;
+            int vardim = (pv->GetDim(6) > 1 ? 1 : 0) + (pv->GetDim(5) > 1 ? 1 : 0) +
+                         (pv->GetDim(4) > 1 ? 1 : 0) + (pv->GetDim(3) > 1 ? 1 : 0) +
+                         (pv->GetDim(2) > 1 ? 1 : 0);
```

Comment:
same for dimension...

Reply:
...and this one

Reply:
Resolving this for now. But I still think it's worth doing down the road.